### DISCUSSION TITLE
React Testing With Deno

### DISCUSSION THREAD

**Discussion by tatemz** [denoland/deno#11952] [Q&A]
It is well-documented how to use React with Deno. However, it is not well-documented on how to go about unit testing components. At first impulse, it seems the simplest solution would be to use the `react-test-renderer`.

However, the `act` method only works if the global variable `__DEV__` is truthy ([link](https://github.com/facebook/react/blob/fc3b6a4118de30ae76f67c2be3d463679c11f4b7/packages/react/src/ReactAct.js#L19))

When `__DEV__` is falsy, `act` outputs the following error message:

```txt
act(...) is not supported in production builds of React, and might not behave as expected.
```

_Note: ⚠️ I don't like global variables. This issue is not advocating to use them, but rather this issue is primarily concerned with healthy testing practices when developing React applications._

How one might go about getting the following test to work?

```tsx
import React from "https://esm.sh/react@17.0.2";
import { act, create } from "https://esm.sh/react-test-renderer@17.0.2";
import { assertNotEquals } from "https://deno.land/std@0.100.0/testing/asserts.ts";

Deno.test({
  name: "react component test",
  fn: async () => {
    const GreeterComponent: React.FC = () => <h1>Hello World</h1>;

    let root: ReturnType<typeof create> = null as unknown as ReturnType<
      typeof create
    >;

    await act(() => {
      root = create(<GreeterComponent />);
    });

    const result = root.toJSON();
    assertNotEquals(result, null);
  },
});
```

**Comment by tatemz** [comment]
Adding some findings. Apparently, ESM.sh has a way to import a dependency as if it was built with the `NODE_ENV` environment variable set to `development`.

https://cdn.esm.sh/#development-mode


```tsx
import "https://esm.sh/raf/polyfill";
import React from "https://esm.sh/react@17.0.2";
import { act, create } from "https://esm.sh/react-test-renderer@17.0.2?dev";
import { assertEquals } from "https://deno.land/std@0.100.0/testing/asserts.ts";

const GreeterComponent: React.FC<{ name: string }> = ({ name }) =>
  <h1>Hello, {name}!</h1>;

Deno.test("react component test", () => {
  const expectedText = "Hello, World!";

  let rendered: ReturnType<typeof create> = null as unknown as ReturnType<
    typeof create
  >;

  act(() => {
    rendered = create(<GreeterComponent name="World" />);
  });

  const h1 = rendered.root.findByType("h1");
  const result = h1.props.children.join("");
  assertEquals(result, expectedText);
});
```

**Reply by tatemz** [reply]
The above test passes, but the `deno test` command does not end. Presumably, there is an unresolved promise somewhere 🤔 

I'd love some ideas on how to get this example working.

**Reply by tatemz** [reply]
This example uses the test renderer but I'd also love to see some ideas on how to use the react dom renderer and `@testing-library/react`

**Reply by caspervonb** [reply]
> The above test passes, but the `deno test` command does not end. Presumably, there is an unresolved promise somewhere 🤔
> 
> I'd love some ideas on how to get this example working.

Try canary, exit condition for tests has changed to "all tests finished" to deal with complex scenarios like these where you don't know what the setup is.
